components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandler.java

```diff
@@ -394,11 +401,46 @@ private boolean validatePKCECode(AuthzCodeDO authzCodeBean, String verificationC
         return true;
     }
 
+    private void revokeAuthorizationCode(AuthzCodeDO authzCodeBean) throws IdentityOAuth2Exception {
+        OAuthTokenPersistenceFactory.getInstance().getAuthorizationCodeDAO().updateAuthorizationCodeState(
+                authzCodeBean.getAuthorizationCode(), OAuthConstants.AuthorizationCodeState.REVOKED);
+        if (log.isDebugEnabled()) {
+            log.debug("Changed state of authorization code : " + authzCodeBean.getAuthorizationCode() + " to revoked");
+        }
+        if (cacheEnabled) {
+            // remove the authorization code from the cache
+            OAuthCache.getInstance().clearCacheEntry(new OAuthCacheKey(
+                    OAuth2Util.buildCacheKeyStringForAuthzCode(authzCodeBean.getConsumerKey(),
+                            authzCodeBean.getAuthorizationCode())));
+            if (log.isDebugEnabled()) {
+                log.debug("Expired Authorization code issued for client " + authzCodeBean.getConsumerKey() +
+                        " was removed from the cache.");
+            }
+        }
+    }
+
     private OAuthAppDO getOAuthAppDO(String clientId) throws IdentityOAuth2Exception {
         try {
             return OAuth2Util.getAppInformationByClientId(clientId);
         } catch (InvalidOAuthClientException e) {
             throw new IdentityOAuth2Exception("Error while retrieving app information for client: " + clientId);
         }
     }
+
+    private void validateRequestParameters(OAuth2AccessTokenReqDTO tokenReq, AuthzCodeDO authzCodeBean)
+            throws IdentityOAuth2Exception {
+        boolean isCallbackValid = false;
+        boolean isPKCEValid = false;
+        try {
+            // If redirect_uri was given in the authorization request,
+            // token request should send matching redirect_uri value
+            isCallbackValid = validateCallbackUrlFromRequest(tokenReq.getCallbackURI(), authzCodeBean.getCallbackUrl());
+            isPKCEValid = validatePKCECode(authzCodeBean, tokenReq.getPkceCodeVerifier());
+        } finally {
+            // In case of exception or invalid request parameters, revoke the authorization code
+            if (!(isCallbackValid && isPKCEValid)) {
```

Comment:
Do we need this if block?
Shouldn't we simply revoke the authorization code regardless of whether the validations passed or failed?